### PR TITLE
[flutter_tools] Generalize waitForExtension

### PR DESCRIPTION
**Pull request by jiahaog**
Genaralizes `waitForExtension` from `resident_devtools_handler.dart` and move it to `vmservice.dart`.

This will allow the same logic to be reused in other places (#76200) where waiting for an extension to be registered is needed.

Slight changes in behavior to make it more general purpose:

- Use `_flutter.listViews` instead to find the Flutter isolate
- Don't wait for `Flutter.FrameworkInitialization`, which might happen before the vmservice is connected and we won't get the message. Let me know if this isn't ok
- Instead listen on the stream of registered extension methods while querying extension methods of the isolate
- Handles the case where the service disappears for a particular device and then skips subsequent service extension calls to the device in `resident_devtools_handler.dart`

### CODE REVIEW

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 1]
```diff
@@ -784,6 +786,37 @@ class FlutterVmService {
...
+            : Completer<void>().future,
+        ),
+        service.onIsolateEvent.firstWhere(
+          (vm_service.Event event) => event.kind == vm_service.EventKind.kServiceExtensionAdded
+              && event.extensionRPC == extensionName,
+        ),
+      ]);
+      return isolateRef;
+    } finally {
+      await service.streamCancel(vm_service.EventStreams.kIsolate);
```
Comment (jiahaog) [inline packages/flutter_tools/lib/src/vmservice.dart:816]: I don't know if this is needed - it seems like it's a common pattern here to *not* cancel streams e.g. https://github.com/flutter/flutter/blob/7672716df06ca58e71dcf7290c2e1286e510bb68/packages/flutter_tools/lib/src/vmservice.dart#L484

I can follow up and add more `streamCancels` elsewhere if that is desired though

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 2]
```diff
@@ -784,6 +786,37 @@ class FlutterVmService {
...
+  /// the stream of added extensions to avoid races.
+  Future<vm_service.IsolateRef> findExtensionIsolate(String extensionName) async {
+    final vm_service.IsolateRef isolateRef = (await getFlutterViews()).first.uiIsolate;
+    try {
+      await service.streamListen(vm_service.EventStreams.kIsolate);
+    } on vm_service.RPCError {
+      // Do nothing, since the tool is already subscribed.
+    }
+
+    try {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/vmservice.dart:801]: I would prefer if this was structured similar to the original code, in a more clearly imperative manner

Comment (jiahaog) [reply packages/flutter_tools/lib/src/vmservice.dart:810]: Done, please take a look again 🙂

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/vmservice.dart:810]: you need to handle the case where isolates is empty. While you are at it, I would consider what happens if there is more than one uiIsolate.

Comment (jiahaog) [reply packages/flutter_tools/lib/src/vmservice.dart:810]: I was under the impression that `getFlutterViews` keeps polling until the isolates are found. Is that not the case here?

Also regarding handling more than one isolate, would it be acceptable to run the querying logic (L813-814) over each `uiIsolate` found?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/vmservice.dart:810]: I believe it will return an empty list if the vm service disconnects.

Comment (jiahaog) [reply packages/flutter_tools/lib/src/vmservice.dart:810]: Sorry, I don't fully understand enough of the VM service to know what to do when it disconnects. Is this a recoverable error where we want to reconnect and make the same call again, or should we fail with an exception?

Comment (jiahaog) [reply packages/flutter_tools/lib/src/vmservice.dart:813]: As discussed over chat, made it throw an exception here, and updated `resident_devtools_handler.dart` to ignore the current `FlutterDevice` when this happens

**Review by jonahwilliams — APPROVED** [review]
LGTM!

**Pull request merged**